NumLib/ODESolver/NonlinearSolver.cpp

```diff
@@ -215,63 +215,65 @@ bool NonlinearSolver<NonlinearSolverTag::Newton>::solve(
         sys.applyKnownSolutionsNewton(J, res, minus_delta_x);
         INFO("[time] Applying Dirichlet BCs took %g s.", time_dirichlet.elapsed());
 
-        if (!sys.isLinear() && _convergence_criterion->hasResidualCheck())
```

Comment:
If the Process is not non-linear the nonlinear solver will automatically stop after one iteration. But it's true, without the residual check the shortcut that you introduced wouldn't be possible.